backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/FunctionReferenceLowering.kt

```diff
@@ -94,6 +94,32 @@ internal class CallableReferenceLowering(val context: Context): FileLoweringPass
                 return result
             }
 
+            // Handle SAM conversions which wrap a function reference:
+            //     class sam$n(private val receiver: R) : Interface { override fun method(...) = receiver.target(...) }
+            //
+            // This is actually very common, as `Interface { something }` is a local function + a SAM-conversion
+            // of a reference to it into an implementation.
+            //
+            // This is needed not only for optimization purposes (producing one less class)
+            // but also for implementing correct equals/hashCode methods.
```

Comment:
This part is not exactly clear.
Does equals/hashCode work correctly for SAM wrappers over anything except function references, i.e. when this optimization is not applied? It should. If it does, then why the optimization here is required for equals/hashCode?

Reply:
Yes, equals/hashCode work correctly for SAM wrappers, but the thing is that if we perform SAM conversion after function references lowering, then for function references to the same function two different implementations will be generated and, hence, references to their invoke functions (which will go into SAM wrappers) will be different as well.

Reply:
What do you mean by "two different implementations will be generated"? What is an implementation here? Could you provide an example?

Reply:
I tried to remove it and all tests still pass, looks like it is not needed (that's weird, because the reason why I added it was falling tests, may be some other things fixed them). Still, as an optimization to not build two wrappers (one over function reference and the other - for SAM conversion) it is still valid.

Reply:
Ah, I figured out why everything works: SAM wrapper in `equals` calls other equals which is `KFunctionImpl.equals', which works correctly in this case.

Reply:
> Still, as an optimization to not build two wrappers (one over function reference and the other - for SAM conversion) it is still valid.

I agree. But this requires different comment.